### PR TITLE
Enhancing returned errors for pkg/kernel.

### PR DESCRIPTION
**Pull request by ybettan**
This is part of the logging enhancement effort which says that all
errors should be wrapped with its context and logging should only be
done at the reconciliation loop level.

Signed-off-by: Yoni Bettan <yonibettan@gmail.com>

### CODE REVIEW

File: pkg/kernel/kernel.go [hunk 1]
```diff
@@ -33,69 +32,60 @@ func (k *kernelData) SetAffineAttributes(obj *unstructured.Unstructured,
...
+	switch obj.GetKind() {
+	case "BuildRun":
 		if err := unstructured.SetNestedField(obj.Object, name, "spec", "buildRef", "name"); err != nil {
-			return err
+			return fmt.Errorf("Could not set %s in %s object: %w", "spec.buildRef.name", "BuildRun", err)
 		}
-	}
-
-	if obj.GetKind() == "DaemonSet" || obj.GetKind() == "Deployment" || obj.GetKind() == "StatefulSet" {
+	case "DaemonSet", "Deployment", "StatefulSet":
```
Comment (ybettan) [inline pkg/kernel/kernel.go:45]: Should it be `Statefulset` instead?

File: pkg/kernel/kernel.go [hunk 2]
```diff
@@ -33,69 +32,60 @@ func (k *kernelData) SetAffineAttributes(obj *unstructured.Unstructured,
...
 	return nil
 }
 
 func (k *kernelData) setVersionNodeAffinity(obj *unstructured.Unstructured, kernelFullVersion string) error {
 
-	if strings.Compare(obj.GetKind(), "DaemonSet") == 0 ||
-		strings.Compare(obj.GetKind(), "Deployment") == 0 ||
-		strings.Compare(obj.GetKind(), "Statefulset") == 0 {
+	switch obj.GetKind() {
+	case "DaemonSet", "Deployment", "Statefulset":
```
Comment (ybettan) [inline pkg/kernel/kernel.go:68]: Should it be `StatefulSet` instead?
On of them is probably bugged.

File: pkg/kernel/kernel.go [hunk 3]
```diff
@@ -33,69 +32,60 @@ func (k *kernelData) SetAffineAttributes(obj *unstructured.Unstructured,
...
 	return nil
 }
 
 func (k *kernelData) setVersionNodeAffinity(obj *unstructured.Unstructured, kernelFullVersion string) error {
 
-	if strings.Compare(obj.GetKind(), "DaemonSet") == 0 ||
-		strings.Compare(obj.GetKind(), "Deployment") == 0 ||
-		strings.Compare(obj.GetKind(), "Statefulset") == 0 {
+	switch obj.GetKind() {
+	case "DaemonSet", "Deployment", "StatefulSet":
```
Comment (ybettan) [inline pkg/kernel/kernel.go:68]: @qbarrand Those should probably be updated as well:
https://github.com/openshift/special-resource-operator/blob/3de8e8fdbc2e39fd2e54c2ebc7bdd20b8efc52e1/internal/resourcehelper/helper.go#L113
https://github.com/openshift/special-resource-operator/blob/3de8e8fdbc2e39fd2e54c2ebc7bdd20b8efc52e1/internal/resourcehelper/helper_test.go#L220

File: pkg/kernel/kernel.go [hunk 4]
```diff
@@ -33,69 +32,60 @@ func (k *kernelData) SetAffineAttributes(obj *unstructured.Unstructured,
...
 		if err := k.versionNodeAffinity(kernelFullVersion, obj, "spec", "nodeSelector"); err != nil {
-			return errors.Wrap(err, "Cannot setup Pod kernel version affinity")
+			return fmt.Errorf("Cannot setup %s's kernel version affinity: %w", obj.GetKind(), err)
 		}
-	}
-	if strings.Compare(obj.GetKind(), "BuildConfig") == 0 {
+	case "BuildConfig":
 		if err := k.versionNodeAffinity(kernelFullVersion, obj, "spec", "nodeSelector"); err != nil {
-			return errors.Wrap(err, "Cannot setup BuildConfig kernel version affinity")
+			return fmt.Errorf("Cannot setup %s's kernel version affinity: %w", obj.GetKind(), err)
```
Comment (yevgeny-shnaidman) [inline pkg/kernel/kernel.go:78]: don't we want to Wrap the error here? it arrives from our function versionNodeAffinity

File: pkg/kernel/kernel.go (hunk 3 above)
Comment (ybettan) [reply pkg/kernel/kernel.go:68]: Yes, I agree it shouldn't be a part of this PR, just mentioned you as I saw the same issue here so we should probably fix that TODO you wrong a while ago.

File: pkg/kernel/kernel_test.go [hunk 5]
```diff
@@ -152,7 +152,7 @@ var _ = Describe("SetVersionNodeAffinity", func() {
 		},
 		Entry("DaemonSet", "DaemonSet"),
 		Entry("Deployment", "Deployment"),
-		Entry("Statefulset", "Statefulset"),
+		Entry("StatefulSet", "StatefulSet"),
```
[the remark this answers is not in the document]
Comment (ybettan) [reply pkg/kernel/kernel_test.go]: Removed.

File: pkg/kernel/kernel.go (hunk 4 above)
Comment (ybettan) [reply pkg/kernel/kernel.go:78]: What @qbarrand said.

File: pkg/kernel/kernel.go [hunk 6]
```diff
@@ -33,69 +32,60 @@ func (k *kernelData) SetAffineAttributes(obj *unstructured.Unstructured,
 	kernelVersion := strings.ReplaceAll(kernelFullVersion, "_", "-")
 	hash64, err := utils.FNV64a(operatingSystemMajorMinor + "-" + kernelVersion)
 	if err != nil {
-		return err
+		return fmt.Errorf("Could not get hash: %w", err)
```
[the remark this answers is not in the document]
Comment (ybettan) [reply pkg/kernel/kernel.go]: Fixed.

File: pkg/kernel/kernel.go [hunk 7]
```diff
@@ -33,69 +32,60 @@ func (k *kernelData) SetAffineAttributes(obj *unstructured.Unstructured,
...
 	}
 	name := obj.GetName() + "-" + hash64
 	obj.SetName(name)
 
-	if obj.GetKind() == "BuildRun" {
+	switch obj.GetKind() {
+	case "BuildRun":
 		if err := unstructured.SetNestedField(obj.Object, name, "spec", "buildRef", "name"); err != nil {
-			return err
+			return fmt.Errorf("could not set %s in %s object: %w", "spec.buildRef.name", "BuildRun", err)
```
[the remark this answers is not in the document]
Comment (ybettan) [reply pkg/kernel/kernel.go]: fixed.

File: pkg/kernel/kernel.go [hunk 8]
```diff
@@ -33,69 +32,60 @@ func (k *kernelData) SetAffineAttributes(obj *unstructured.Unstructured,
...
 		}
-	}
-
-	if obj.GetKind() == "DaemonSet" || obj.GetKind() == "Deployment" || obj.GetKind() == "StatefulSet" {
+	case "DaemonSet", "Deployment", "StatefulSet":
 		if err := unstructured.SetNestedField(obj.Object, name, "metadata", "labels", "app"); err != nil {
-			return err
+			return fmt.Errorf("could not set %s in %s object: %w", "metadata.labels.app", "DaemonSet", err)
```
[the remark this answers is not in the document]
Comment (ybettan) [reply pkg/kernel/kernel.go]: Fixed.

File: pkg/kernel/kernel.go [hunk 9]
```diff
@@ -33,69 +32,60 @@ func (k *kernelData) SetAffineAttributes(obj *unstructured.Unstructured,
...
 		}
 
 		if err := unstructured.SetNestedField(obj.Object, name, "spec", "selector", "matchLabels", "app"); err != nil {
-			return err
-		}
-
-		if err := unstructured.SetNestedField(obj.Object, name, "spec", "template", "metadata", "labels", "app"); err != nil {
-			return err
+			return fmt.Errorf("could not set %s in %s object: %w", "spec.selector.matchLabels.app", "DaemonSet", err)
```
[the remark this answers is not in the document]
Comment (ybettan) [reply pkg/kernel/kernel.go]: Fixed.

File: pkg/kernel/kernel.go [hunk 10]
```diff
@@ -33,69 +32,60 @@ func (k *kernelData) SetAffineAttributes(obj *unstructured.Unstructured,
...
 		}
 
 		if err := unstructured.SetNestedField(obj.Object, name, "spec", "template", "metadata", "labels", "app"); err != nil {
-			return err
+			return fmt.Errorf("could not set %s in %s object: %w", "spec.template.metadata.labels.app", "DaemonSet", err)
```
[the remark this answers is not in the document]
Comment (ybettan) [reply pkg/kernel/kernel.go]: Fixed.

**Pull request merged**